File: Windows/CharacterAppearance.cs
using System.Drawing;

namespace PCNoShitaNoChikaramochi;

public enum Gender { Male, Female }
public enum SkinTone { Porcelain, Light, Medium, Tan, Deep }
public enum ShirtColor { Red, Blue, Green, Purple, Orange, Black, White, Pink }

public record struct AppearanceSnapshot(Gender Gender, SkinTone SkinTone, ShirtColor ShirtColor);

public static class AppearanceColors
{
    public static Color Skin(SkinTone t) => t switch
    {
        SkinTone.Porcelain => Color.FromArgb(255, 235, 215),
        SkinTone.Light     => Color.FromArgb(255, 214, 168),
        SkinTone.Medium    => Color.FromArgb(219, 168, 128),
        SkinTone.Tan       => Color.FromArgb(179, 128,  89),
        SkinTone.Deep      => Color.FromArgb(107,  69,  46),
        _ => Color.Beige
    };

    public static Color SkinShade(SkinTone t) => t switch
    {
        SkinTone.Porcelain => Color.FromArgb(224, 189, 158),
        SkinTone.Light     => Color.FromArgb(199, 148, 107),
        SkinTone.Medium    => Color.FromArgb(158, 107,  71),
        SkinTone.Tan       => Color.FromArgb(122,  77,  46),
        SkinTone.Deep      => Color.FromArgb( 66,  38,  20),
        _ => Color.SaddleBrown
    };

    public static Color Shirt(ShirtColor c) => c switch
    {
        ShirtColor.Red    => Color.FromArgb(214, 41, 46),
        ShirtColor.Blue   => Color.FromArgb(46, 115, 217),
        ShirtColor.Green  => Color.FromArgb(56, 153, 89),
        ShirtColor.Purple => Color.FromArgb(128, 64, 191),
        ShirtColor.Orange => Color.FromArgb(242, 140, 46),
        ShirtColor.Black  => Color.FromArgb(38, 38, 43),
        ShirtColor.White  => Color.FromArgb(242, 242, 247),
        ShirtColor.Pink   => Color.FromArgb(242, 140, 179),
        _ => Color.Red
    };

    public static Color Plate(CharacterState s) => s switch
    {
        CharacterState.VeryEasy or CharacterState.Easy => Color.FromArgb(76, 166, 242),  // 軽い: 青
        CharacterState.Normal     => Color.FromArgb(102, 191, 102),                       // 中: 緑
        CharacterState.Heavy      => Color.FromArgb(242, 140, 51),                        // 重い: 橙
        CharacterState.Struggling => Color.FromArgb(230, 64, 64),                         // 超重: 赤
        CharacterState.Crushed    => Color.FromArgb(140, 26, 26),                         // 致死量: 暗赤
        _ => Color.Gray
    };

    public static Color Hair => Color.FromArgb(46, 31, 26);
    public static Color Shorts => Color.FromArgb(46, 46, 56);
    public static Color Bar => Color.FromArgb(199, 199, 209);

    public static string LabelOf(Gender g) => g switch
    {
        Gender.Male => "男性",
        Gender.Female => "女性",
        _ => "?"
    };

    public static string LabelOf(SkinTone t) => t switch
    {
        SkinTone.Porcelain => "とても明るい",
        SkinTone.Light     => "明るい",
        SkinTone.Medium    => "中間",
        SkinTone.Tan       => "小麦",
        SkinTone.Deep      => "濃い",
        _ => "?"
    };

    public static string LabelOf(ShirtColor c) => c switch
    {
        ShirtColor.Red    => "赤",
        ShirtColor.Blue   => "青",
        ShirtColor.Green  => "緑",
        ShirtColor.Purple => "紫",
        ShirtColor.Orange => "橙",
        ShirtColor.Black  => "黒",
        ShirtColor.White  => "白",
        ShirtColor.Pink   => "桃",
        _ => "?"
    };
}

File: Sources/PCNoShitaNoChikaramochi/CharacterAppearance.swift
import Foundation
import SwiftUI

enum Gender: String, CaseIterable, Identifiable {
    case male
    case female

    var id: String { rawValue }

    var label: String {
        switch self {
        case .male: return "男性"
        case .female: return "女性"
        }
    }
}

enum SkinTone: String, CaseIterable, Identifiable {
    case porcelain
    case light
    case medium
    case tan
    case deep

    var id: String { rawValue }

    var label: String {
        switch self {
        case .porcelain: return "とても明るい"
        case .light: return "明るい"
        case .medium: return "中間"
        case .tan: return "小麦"
        case .deep: return "濃い"
        }
    }

    var primary: Color {
        switch self {
        case .porcelain: return Color(red: 1.00, green: 0.92, blue: 0.84)
        case .light: return Color(red: 1.00, green: 0.84, blue: 0.66)
        case .medium: return Color(red: 0.86, green: 0.66, blue: 0.50)
        case .tan: return Color(red: 0.70, green: 0.50, blue: 0.35)
        case .deep: return Color(red: 0.42, green: 0.27, blue: 0.18)
        }
    }

    var shade: Color {
        switch self {
        case .porcelain: return Color(red: 0.88, green: 0.74, blue: 0.62)
        case .light: return Color(red: 0.78, green: 0.58, blue: 0.42)
        case .medium: return Color(red: 0.62, green: 0.42, blue: 0.28)
        case .tan: return Color(red: 0.48, green: 0.30, blue: 0.18)
        case .deep: return Color(red: 0.26, green: 0.15, blue: 0.08)
        }
    }
}

enum ShirtColor: String, CaseIterable, Identifiable {
    case red
    case blue
    case green
    case purple
    case orange
    case black
    case white
    case pink

    var id: String { rawValue }

    var label: String {
        switch self {
        case .red: return "赤"
        case .blue: return "青"
        case .green: return "緑"
        case .purple: return "紫"
        case .orange: return "橙"
        case .black: return "黒"
        case .white: return "白"
        case .pink: return "桃"
        }
    }

    var primary: Color {
        switch self {
        case .red: return Color(red: 0.84, green: 0.16, blue: 0.18)
        case .blue: return Color(red: 0.18, green: 0.45, blue: 0.85)
        case .green: return Color(red: 0.22, green: 0.60, blue: 0.35)
        case .purple: return Color(red: 0.50, green: 0.25, blue: 0.75)
        case .orange: return Color(red: 0.95, green: 0.55, blue: 0.18)
        case .black: return Color(red: 0.15, green: 0.15, blue: 0.17)
        case .white: return Color(red: 0.95, green: 0.95, blue: 0.97)
        case .pink: return Color(red: 0.95, green: 0.55, blue: 0.70)
        }
    }
}

struct AppearanceSnapshot: Equatable {
    var gender: Gender
    var skinTone: SkinTone
    var shirtColor: ShirtColor
}

@MainActor
final class AppearanceStore: ObservableObject {
    @Published var gender: Gender {
        didSet { persist() }
    }
    @Published var skinTone: SkinTone {
        didSet { persist() }
    }
    @Published var shirtColor: ShirtColor {
        didSet { persist() }
    }

    static let keyGender = "appearance.gender"
    static let keySkin = "appearance.skin"
    static let keyShirt = "appearance.shirt"

    private let defaults: UserDefaults

    /// - Parameter defaults: 永続化先。テストでは独立した suite を渡せる。
    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.gender = Gender(rawValue: defaults.string(forKey: Self.keyGender) ?? "") ?? .male
        self.skinTone = SkinTone(rawValue: defaults.string(forKey: Self.keySkin) ?? "") ?? .light
        self.shirtColor = ShirtColor(rawValue: defaults.string(forKey: Self.keyShirt) ?? "") ?? .red
    }

    var snapshot: AppearanceSnapshot {
        AppearanceSnapshot(gender: gender, skinTone: skinTone, shirtColor: shirtColor)
    }

    private func persist() {
        defaults.set(gender.rawValue, forKey: Self.keyGender)
        defaults.set(skinTone.rawValue, forKey: Self.keySkin)
        defaults.set(shirtColor.rawValue, forKey: Self.keyShirt)
    }
}
